README.md
# customs-request

## Installation

### Using npm:

```bash
$ npm install customs-request
```

## Configuration

- ### main_dir `string`
  Директория относительно которой будут находиться все файлы проекта, корень проекта: картинки, шаблоны, страницы и т.д. - _задается абсолютным путем_.

- ### path_to_error_pages_dir `string`
  Путь к директории страниц ошибок имя страницы должно соответствовать коду ошибки. Имеет приоритет над
  path_to_error_agent - _задается абсолютным путем_.
    - 404.html
    - 503.html
    - 401.html
    - ...


- ### path_to_error_agent `string`
  Путь к error_agent - это скрипт, который отправляется в HTML контейнере, в случае возникновения ошибок 401, 404, 500
  ... и т.п. . Будет запущен на клиенте, для скрипта подготовлен
  объект: `window.error_date = {status_code, status_message, url}`, является альтернативой для path_to_error_pages_dir и
  игнорируется в случае если path_to_error_pages_dir задан - _задается абсолютным путем_.


- ### path_to_app_dir `string`
  Путь к директории с файлами пользовательских приложений
    - Уровень доступа к приложению или конкретному методу приложения устанавливается в конфиге.
    - Вызов метода: `/app_name/method_name?`.

- ### urls `array`
  Указать можно путь к директории, файлу или имя приложению и его метод. Ограничение для директории/приложения
  распространяется на все вложенные файлы и директории/методы, при условии, если вложенный файл или директория/метод не
  указаны отдельно, в таком случае действует указанные ограничение.
    - `objekt`
        - **value** `string`  Устанавливает путь относительно корня _main_dir_ к приложению | методу приложения | директории | файлу.
        - **access_level** **|** **access_level_only** `number` Устанавливает уровень доступа к приложению | методу приложения | директории | 
          файлу относительно _main_dir_. Всего 4 уровня: 
          main admin - 0, admin - 1, user -
          2, не подтвержденный user - 3. Если **access_level**: урл будет доступен от указанного до наивысшего.
          Если **access_level_only**: Доступно только для указанного уровня. Если **access_level_only** и **access_level** не указан доступно для всех посетителей.
        - **app** `boolean` является ли приложением, если false или не указан, значит это путь к директории/файлу

- ### db `object`
    представление базы данных, объект с методами:
    - `register_user()`
        - Аргументы:
            - `object`
                - email* `string`
                - password*  `string`,
                - token* `string`,
                - user_level* `string`,
                - ... ( other_field)`string`,
        - Возвращает:
            - `string` id юзера в случае успеха
            - **ИЛИ**
            - `boolean` false в противном случае
              <br><br/>
    - `update_user_field()`
        - Аргументы:
            - field_name* `string`
            - new_value*  `string`,
            - user_id* `string`,
        - Возвращает:
            - `boolean`
                - **true** поле в таблице users обновлено успешно
                - **false** ошибка обновления
                  <br><br/>
    - `get_users()`
        - Возвращает:
            - `array` массив строк из таблицы users
              - `object`
                  - email* `string`
                  - password*  `string`,
                  - token* `string`,
                  - user_level* `string`,
                  - ... ( other_field)`string`, 
            - **ИЛИ**
            - `boolean` false в противном случае

### Пример файла config

```js
// server.config.js

const path = require("path");
const db = require('./db'); //Объект с метадами для записи и чтения таблицы users в БД.
const config = {
    main_dir: path.join(__dirname, '/src/assets'), // main project directory. Use full path - __dirname, 'current_dir'
    error_pages_dir: path.join(__dirname, '/src/assets/pages_errors'), // directory with error pages: 404.html, 401.html, ..., .
    urls: [
        {
            value: '/sys_app',
            access_level: 1,
            app: true
            // Доступны все методы приложения sys_app для пользователей с уровнем 1, 0 (admin, main_admin)
        },
        {
            value: '/sys_app/list_users',
            access_level: 2,
            app: true
            // Метод list_users приложения sys_app доступен  пользователей с уровнем 2, 1, 0 (user, admin, main_admin)
        },
        {
            value: '/users',
            access_level_only: 2,
            app: true
            // Директория users доступно только level 2 (user)
        },
        {
            value: '/users/img_publick',
            app: true
            // Директория users доступно только level 2 (user), но под дириктория img_publick доступна всем 
        },
        {
            value: '/users/statistic.json',
            access_level: 1,
            app: true
            // Директория users доступно только level 2 (user), 
            // но файл /users/statistic.json' доступен level 1, 0 (admin, main_admin)
        },
        {
            value: '/free_app',
            app: true,
            // Приложение free_app доступно всем
        },
        {
            value: '/free_app/num_call',
            access_level: 1,
            app: true,
            // Приложение free_app доступно всем, но его метод num_call доступен только 1 и 0 - admin, main_admin
        },
    ],
    path_to_app_dir: path.join(__dirname, './apps'),
    db

    // db.register_user()
    // db.update_user_field()
    // db.get_users()
}

module.exports = config


```

## IDENT API
Встроенное приложение регистрации и авторизации 
### Регистрация пользователя

**Метод POST:** `/ident/register_user?`

```js
//Клиент для регистрации должен придоставить форму с обязательными полями:

body = {
    email: 'name@email.com',
    password: '123_password',
    answer_antibot: "Фраза отвте на антибот",
    id_task_antibot: "KYT76K",
}

response = "ок"
```

**Порядок:**

- Клиент
    1) запрашивает задание антибот или пользуется ранее запрошенным (у каждого решения **два** использования)
    2) представляет пустую форму пользователю для регистрации
    3) после ввода отправляет запрос на сервер, ответ:
        - 200 ок - все ок
        - 401 email_taken _- занят уже_
        - 401 email_syntax_error _- не правильный синтаксис у e-mail_
        - 401 password_syntax_error _- не правильный синтаксис у password_
        - 401 check_bot_error _- не прошел проверку на бот_

### Авторизация пользователя

**Метод POST:** `/ident/auth_user?`

```js
//Клиент для авторизации должен придоставить форму с полями:

body = {
    email: 'name@email.com',
    password: '123_password',
    answer_antibot: "Фраза отвте на антибот",
    id_task_antibot: "KYT76K",
}

response = "ок"
```

**Порядок:**

- Клиент
    1) запрашивает задание антибот или пользуется ранее запрошенным (у каждого решения **два** использования)
    2) представляет пустую форму пользователю для авторизации
    3) после ввода отправляет запрос на сервер, ответ:
        - 200 ок - все ок установлен новый токен
        - 401 auth_error _- не верный логи или пароль_
        - 401 check_bot_error _- не прошел проверку на бот_

### Проверка e-mail

**Метод GET:** `/ident/exist_user?`

```js
//Клиент для авторизации должен придоставить форму с полями:

body = {
    email: 'name@email.com',
    answer_antibot: "Фраза отвте на антибот",
    id_task: "KYT76K",
}
```

**Порядок:**

- Клиент
    1) запрашивает задание антибот или пользуется ранее запрошенным (у каждого решения **два** использования)
    2) представляет пустую форму пользователю для авторизации
    3) после ввода отправляет запрос на сервер, ответ:

        - 200 ок - все ок
        - 401 email_taken _- занят уже_
        - 401 email_syntax_error _- не правильный синтаксис у e-mail_
        - 401 password_syntax_error _- не правильный синтаксис у password_
        - 401 check_bot_error _- не прошел проверку на бот_

### Получить задания для антибот

**Метод GET:** `/ident/get_antibot_task?`

```js
// Сервер присылает задачку на антибот в виде ссылки н картинку и номер задачи, ответ можно использовать два раза:

response = {
    task: "ссылка на картинку с заданием",
    id_task: "KYT76K",
}
```

**Порядок:**

- Клиент запрашивает задание антибот (у каждого решения **два** использования):
    - 200 ок _- все ок_

## Usage

```js

// server.js

const Client = require('customs-request');
const config = require('./config');
Client.set_config(config);

//...

async function handlerServer(request, response) {
    let client = new Client(request, response);
    await client.init();
}

//...
```

## Data Base

### users

- id
- email
- password
- token
- level

## Customs API

Ваши API должны находится в директории указанной в **config.path_to_app_dir**. Имя файла это имя API, и метод
соответственно имя метода по которым клиент будет вызывать ваше API. Уровни доступа необходимо прописать в config.

### Аргумент для методов ваших API:

- `object`
    - user `object`
        - user_id* `string`,
        - user_level* `string`,
    - body `object`
    - query `object`
    - req `object: node.js<http.IncomingMessage>`
    - res `object: node.js<http.ServerResponse>`
    - url_level `string`
    - url_level_only `string`
    - url_value `string`
    - apps `object`
    - send `function`

### Custom метод должен Возвращать:

Статус код и статус сообщение уже определены, так же их можно изменить через response, кроме это можно отправить ответ
или просто вернуть его из метода. Если метод ни чего не send и не return, тогда, code 200, send пустая строка

- `string(JSON)`: Данные, для отправки клиенту

## License

[MIT](LICENSE)

CHANGELOG.md

ereer - `register_user()`
            - Аргументы:
                - `object`
                    - email* `string`
                    - password*  `string`,
                    - token* `string`,
                    - user_level* `string`,
                    - ... ( other_field)`string`,

            - Возвращает:
                - `string` id юзера в случае успеха
                    - ИЛИ
                - `boolean` false в противном случае
                  <br><br/>
        - `update_user_field()`
            - Аргументы:
                - field_name* `string`
                - new_value*  `string`,
                - user_id* `string`,

            - Возвращает:
                - `boolean`
                    - **true** поле в таблице users обновлено успешно
                    - **false** ошибка обновления
                      <br><br/>
        - `get_users()`
            - Возвращает:
                - `array` массив строк из таблицы users
                    - `object`
                        - email* `string`
                        - password*  `string`,
                        - token* `string`,
                        - user_level* `string`,
                        - ... ( other_field)`string`,
                - **ИЛИ**
                - `boolean` false в противном случае


lib/identify.js
const get_token = require("get_token_char");
const crypto = require('crypto');
const {raw} = require("express");
const RotaryCaptcha = require('../../rotary_captcha');
const captcha = new RotaryCaptcha();

const store_bot = {}

let users,
    token_name,
    db;


function _check_answer_antibot(id_task, answer) {
    console.log(id_task, answer);
    return captcha.check_task(id_task, answer);
}

//  hash_password(password) --> sha256(password + secret_key)
async function hash_password(password) {
    if (typeof password !== 'string' || !password.length) {
        console.warn(`hash_password(): bad arg password = ${password}`);
        return;
    }
    let key = 'dLdK8Lf0KOrLKj77jhKKMaqZqAdsWplMghDkmE';
    return crypto
        .createHmac('sha256', key)
        .update(password)
        .digest('hex');
}

async function _check_password(email, password) {
    password = await hash_password(password);
    let user_id = await db.check_email_password(email, password);
    console.log(user_id, 37);
    if (user_id) {
        let new_token = get_token();
        for (let i = 0; i < users.length; i++) {
            let user = users[i];
            if (user_id === user.id) {
                user.token = new_token;
                return {user_id, new_token}
            }
        }
    }
    return false
}

function _is_email_busy(email) {
    if (typeof email !== "string") return false;
    //return await db.is_email_busy();
    return users.some(user => user.email.toLowerCase() === email.toLowerCase());
}

function _check_format_email(email) {
    if (typeof email === "string") {
        return email.toLowerCase();
    }
    return false;
}

// check_user_token(token) --> {user_id, asses_level} | false
async function check_user_token(token) {
    if (!Array.from(users)) return;
    let res = false;
    users.some(user => {
        if (user.token === token) {
            res = {user_id: user.id, asses_level: user.asses_level}
            return true
        }
    });
    return res
}

async function add_user(new_user) {
    let {id} = await db.register_user(new_user);
    if (id) {
        new_user.id = id;
        users.push(new_user);
        return id
    }
    return false
}

function error_valid_register_data(email, password) {
    return false
}

function include_url(config) {
    config.urls = [
        {
            value: '/ident',
            app: true
        },
        {
            value: '/get_captcha',
            app: true
        },
        {
            value: '/ident/_is_user',
            app: true,
            access_level: 0
        },
        {
            value: '/ident/_hash_password',
            app: true,
            access_level: 0
        },
        {
            value: '/ident/_get_token',
            app: true,
            access_level: 0
        },
        {
            value: '/ident/_local_register_user',
            app: true,
            access_level: 0
        },
    ].concat(config.urls)
}

function get_incomming_data(client) {
    let data = false;
    let {body, query, fields, method_html} = client;

    if (method_html.toLowerCase() === 'get') data = query;
    else if (method_html.toLowerCase() === 'post') {
        if (Object.values(body).length) data = body;
        data = fields
    }
    return data;
}

class Ident {
    constructor(config) {
        users = config.users;
        db = config.db;
        token_name = config.token_name;
        include_url(config);
    }

    async get_captcha() {
        let task = await captcha.create_task()
        return JSON.stringify(task);
    }

    _hash_password = hash_password;

    async _is_user(client) {
        let {cookie, send} = client;
        if (cookie && cookie[token_name]) {  // return {user_id, asses_level} | false
            return await check_user_token(cookie[token_name]);
        }
        return false;
    }

    async is_user(client) {
        let {send} = client;
        let res = await this._is_user(client);
        if (res) {
            send(200, 'auth', 'auth', false);
            return;
        }
        send(401, 'not_auth', 'not_auth', false);
    }

    async _get_token(user_id) {
        for (let i = 0; i < users.length; i++) {
            let user = users[i];
            if (user_id === user.id) {
                return user.token;
            }
        }
        return  false
    }

    async _local_register_user(client) {
        let {send, method_html} = client;
        let data = get_incomming_data(client);
        if (!data) {
            console.warn(`html method incorrect ${method_html}`);
            send(401, 'not_date_for_register', 'not_date_for_register', false);
            return false;
        }
        if (!data.password || !data.email) {
            console.warn('email && password undefined');
            send(401, 'not_all_filled', 'not_all_filled', false);
            return false;
        }
        let {email, password, id_task_antibot, answer_antibot, nickname} = data;

        if (!_check_answer_antibot(id_task_antibot, answer_antibot)) {
            send(401, 'check_bot_error', 'check_bot_error', false);
            return false;
        }

        if (_is_email_busy(email)) {
            send(401, 'email_taken', 'email_taken', false)
            return false;
        }

        email = _check_format_email(email)
        if (!email) {
            send(401, 'bad_email_format', 'bad_email_format', false)
            return false;
        }
        password = await hash_password(password);
        if (!password) {
            send(401, 'data_incorrect', 'data_incorrect', false); //send(401, 'bad_password_for_register', 'bad_password_for_register', false)
            return false;
        }
        let new_user = {
            email,
            password,
            token: get_token(),
            asses_level: 3,
            nickname
        };
        if (
            !new_user ||
            !new_user.email ||
            !new_user.password ||
            !new_user.token ||
            !new_user.nickname ||
            !new_user.asses_level
        ) {
            send(401, 'data_incorrect', 'data_incorrect', false);
        }
        let id = await add_user(new_user);
        if (id) return id;
        send(500, 'no_solution_found_during_registration', 'no_solution_found_during_registration', false);
        return false;
    }

    async register_user(client) {
        let {send} = client;
        let id = await this._local_register_user(client);
        if (id) send(200, 'successful_registration', 'successful_registration', false);
    }

    async log_out(client) {
        let {send, res} = client;
        res.cookie([token_name], 'none', {
            expires: new Date(Date.now() - 1000),
            httpOnly: true
        });
        send(200, 'log_out', 'log_out', false); //(code, message, data, send_agent)
    }

    async auth_user(client) {
        let {send, res, method_html} = client;
        let data = get_incomming_data(client);
        if (!data) {
            console.warn(`html method incorrect ${method_html}`);
            return send(401, 'not_date_for_register', 'not_date_for_register', false);
        }
        let {email, password, id_task_antibot, answer_antibot} = data;
        if (!_check_answer_antibot(id_task_antibot, answer_antibot)) {
            send(401, 'check_bot_error', 'check_bot_error', false);
            return;
        }
        email = _check_format_email(email)
        if (!email) {
            send(401, 'bad_email_format', 'bad_email_format', false)
            return false;
        }
        let {user_id, new_token} = await _check_password(email, password); // {user_id, new_token} | false
        if (user_id) {
            let upd = await db.update_user_field('token', new_token, user_id);
            if (upd) {
                res.cookie([token_name], new_token, {
                    expires: new Date(Date.now() + 2592e+9),
                    httpOnly: true
                });
                client.send(200, 'successful_registration', 'successful_registration', false); //(code, message, data, send_agent)
            }
        }
        send(401, 'data_incorrect', 'data_incorrect', false);
    }

    async is_email_busy(client) {
        let data = get_incomming_data(client);
        let {send} = client;
        if (!data) {
            console.warn(`html method incorrect ${method_html}`);
            return send(401, 'not_date_for_register');
        }
        let {email} = data;

        if (_is_email_busy(email)) {
            send(401, 'email_taken')
            return;
        }
        return 'ok'
    }
}

module.exports = function (config) {
    return new Ident(config)
}

